### PR TITLE
[gait-selection-develop] Feature: subgait transitions

### PR DESCRIPTION
**Pull request by Olavhaasie**
Closes PM-543

## Description
This PR adds the final feature to the new gait selection: Transition gaits :tada: The only small thing that is missing is that you can start the walk gait from a small or large size (PM-554)

I have implemented the transitions on subgait level and are therefore included in the gait graph.

## Changes
* Add increase and decrease size transitions in state machine
* Implement transitioning on `SetpointsGait`
* Remove `GaitSelection` dependency on `TransitionSubgait` and create from subgait objects instead
* Added forgotten timestamp to published messages

## Testing
Checkout `gait-selection-develop` on gait-files, monitor and simulation. Run the simulation and make sure you use `gait_directory:=minimal`. Then start walking and transitioning between sizes.

### CODE REVIEW

File: march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py [hunk 1]
```diff
@@ -53,32 +57,60 @@ def final_position(self):
...
-            trajectory = self.subgaits[next_subgait].to_joint_trajectory_msg()
-            self._current_subgait = next_subgait
+            self._current_subgait = self.subgaits[next_subgait]
+            trajectory = self._current_subgait.to_joint_trajectory_msg()
             self._time_since_start = 0.0
             return trajectory, False
 
+    def transition(self, transition_request):
+        if transition_request == TransitionRequest.DECREASE_SIZE:
+            name = self.graph[(self._current_subgait.subgait_name, self.graph.DECREASE_SIZE)]
```
Comment (Roelemans) [inline march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py:103]: Feels a bit strange to basically use two different Enums here, but not sure if there is a better way

**Review by Roelemans — CHANGES_REQUESTED** [review]
image
When you click "increase step size" during a transition, the simulation crashes :crying_cat_face: .

File: march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py [hunk 2]
```diff
@@ -53,32 +57,60 @@ def final_position(self):
...
+                    s=self._transition_to_subgait.subgait_name))
+                self._current_subgait = transition_subgait
+                self._time_since_start = 0.0
+                self._is_transitioning = True
+                return transition_subgait.to_joint_trajectory_msg(), False
+            elif self._transition_to_subgait is not None and self._is_transitioning:
+                next_subgait = self._transition_to_subgait.subgait_name
+                self._transition_to_subgait = None
+                self._is_transitioning = False
+            elif self._should_stop:
```
[the remark this answers is not in the document]
Comment (Olavhaasie) [reply march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py:85]: The first one I can easily solve indeed. The second one is a lot harder to execute, I could create an issue for the next team :innocent:

File: march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py (hunk 1 above)
Comment (Olavhaasie) [reply march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py:103]: Yes I agree, it does not look natural. However, these represent two different things. The `TransitionRequest` is an actual enum with different values, whereas the `graph.DECREASE_SIZE` just a string value is for the name of the transition.

**Comment by Olavhaasie** [comment]
> When you click "increase step size" during a transition, the simulation crashes crying_cat_face .

Thanks for your review and for finding the crashes. I fixed it so that you can not execute stop or transition during the other.

File: march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Roelemans) [reply march_gait_selection/src/march_gait_selection/state_machine/setpoints_gait.py]: Lekker :ostrich: :stuck_out_tongue:

**Review by Roelemans — CHANGES_REQUESTED** [review]
Good job fixing the errors! One last thing: when pressing stop during a transition now, the stop now gets rejected instead of postponed until after the transition. This is arguably worse since you would prefer the stop to happen as quickly as possible even if you can't do it immediately. However, you can also just postpone this to the other issue :innocent: 
image

**Comment by Olavhaasie** [comment]
> However, you can also just postpone this to the other issue

That's exactly what I was thinking :stuck_out_tongue_winking_eye:

**Review by Roelemans — APPROVED** [review]

**Review by JorisWeeda — CHANGES_REQUESTED** [review]
I get an error when building with your testing settings;
`
--- stderr: march_gait_files
CMake Error at cmake_install.cmake:56 (file):
  file INSTALL cannot find
  "/home/joris/march_ws/src/gait-files/march_gait_files/airgait".
`
Looking at your cmake file u state that there is an _airgait_ folder but this does not exist. Not sure why Roel did not encounter this problem tho.

**Comment by Olavhaasie** [comment]
> I get an error when building with your testing settings;
> `--- stderr: march_gait_files CMake Error at cmake_install.cmake:56 (file): file INSTALL cannot find "/home/joris/march_ws/src/gait-files/march_gait_files/airgait".`
> Looking at your cmake file u state that there is an _airgait_ folder but this does not exist. Not sure why Roel did not encounter this problem tho.

Oopsie, I fixed it. I did this after Roel's review

**Review by JorisWeeda — APPROVED** [review]
Looks nice, except for the _stop_ problem as roel described everything looks fine by me. nice work! :fire:

**Pull request merged**